### PR TITLE
Automatically update dom0 and VM configs over time

### LINKED ISSUE #24: Workstation update process
If we decide to make a change to the VM structure here, then we'd need to ship an update (if the hot take in #17 pans out, then via the `securedrop-workstation` `dom0` package). Without further modifications in Qubes, they'd be notified of a `dom0` update and we'd need to manually have users kick off an update in `dom0`. 

Is this what we want? I think ideally we don't want to require user intervention to update the `securedrop-workstation` config.

### PR DESCRIPTION
**Pull request by conorsch**
### Overview

Installs a new script to dom0 called `securedrop-update`, and symlinks it into cron.daily, so it runs regularly without manual activation. The script will:

* pull in all new `dom0` updates
* start each TemplateVM and update its packages internally (included apt-test-qubes deb packages)
* display clunky feedback about the update process with a pretty SD logo

Closes #24.

### Screenshots

Here's what the tool looks like in action:

sd-update-2

### Metrics

The script takes ~10m to run. (If you haven't updated your TemplateVMs in a while, then it could take longer, ~30m, the first time you run it!) It will take less if you don't currently have the Workstation VMs configured (e.g. if you've recently run `make clean`).

```
[user@dom0 securedrop-workstation]$ time sudo securedrop-update

[...snip...]

local:
    ----------
whonix-gw-14: OK
fedora-28: OK
openvpn: OK
sd-svs-disp-template: OK
signal-template: OK
debian-9: OK
whonix-ws-14: OK
sd-svs-template: OK
sd-journalist-template: OK
sd-whonix-template: OK
sd-workstation-template: OK
securedrop-workstation: OK

real	10m3.111s
user	0m37.405s
sys	1m1.834s
```

### Next steps
As with other dom0-related configs, we'll need to package these as an RPM (#171) eventually.

### CODE REVIEW

File: dom0/securedrop-update [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# but we *first* want the freshest RPMs from dom0, *then* we'll want to
+# update the VMs themselves.
+securedrop-update-feedback "SecureDrop: Updating dom0 configuration..."
+sudo qubes-dom0-update -y
+
+securedrop-update-feedback "SecureDrop: Updating application..."
+qubesctl --templates \
+    --max-concurrency "$SECUREDROP_MAX_CONCURRENCY" \
+    pkg.upgrade refresh=true
+
```
Comment (emkll) [inline dom0/securedrop-update]:
```suggestion

securedrop-update-feedback "SecureDrop: Updating TorBrowser..."
qvm-run sd-journalist-template "update-torbrowser --noask"
qvm-shutdown sd-journalist-template
```

**Comment by conorsch** [comment]
@emkll Does the `update-tor` command resolve #148? If so, might be best to submit as a follow-up PR... I gather the appropriate testing workflow for resolution there would be to uninstall the Whonix templates from dom0, then re-apply. Agree?

File: dom0/securedrop-update [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Display GUI feedback about update process
+function securedrop-update-feedback() {
+    # Unpack msg as arg1
+    local msg="$1"
+    shift
+
+    # Running `notify-send` as root doesn't work, must be normal user.
+    # Setting 30s expire time (in ms) since it's a long-running cmd.
+    su user -c "notify-send \
```
[the remark this answers is not in the document]
Comment (conorsch) [reply dom0/securedrop-update]: Ouch, that's a good flag. We still need to drop privileges in here; or we could dig more in the `notify-send` settings. Off the cuff, inspecting `/home/` for a single dirname _should_ give us whatever the name of the (single) custom user is. Make sense?

[the remark this answers is not in the document]
Comment (conorsch) [reply dom0/securedrop-update]: We can safely assume that the normal user configured at install time has uid 1000; so:

```
id -nu 1000
```

Then we `su` to that user to run the `notify-send` commands.

**Comment by conorsch** [comment]
Discussed review changes with @emkll; to summarize, next steps for this PR are:

* [ ] implement custom "user" handling for notify-send command
* [ ] use `state.highstate` to enforce VM settings (not just packages inside VMs)
* [ ] drop `--templates` flags and `qubes-dom0-update` cmd (as @marmarek helpfully pointed out is superfluous here)
* [ ] add a GUI notification recommending reboot of the full workstation after updates are applied.

The GUI notification recommending a reboot has UX implications; but it's better to warn than reboot the AppVMs while they may be in use. After this PR is stable and merged, we can ping @ninavizz (and @eloquence) for assistance evaluating the UX impact and recommended improvements.

For the `update-tor` fix, we'll handle separate, in a discrete PR targeting #148.

**Comment by conorsch** [comment]
All outstanding items addressed. Please have another look, @emkll! Would love your thoughts as well, @joshuathayer .

File: dom0/securedrop-update [hunk 3]
```diff
@@ -21,26 +21,34 @@ function securedrop-update-feedback() {
...
 # `qubesctl pkg.upgrade` will automatically update dom0 packages, as well,
 # but we *first* want the freshest RPMs from dom0, *then* we'll want to
 # update the VMs themselves.
-securedrop-update-feedback "SecureDrop: Updating dom0 configuration..."
-sudo qubes-dom0-update -y
 
-securedrop-update-feedback "SecureDrop: Updating application..."
-qubesctl --templates \
+securedrop-update-feedback "Updating application..."
+qubesctl \
```
Comment (marmarek) [inline dom0/securedrop-update]: Oh, sorry, I think I wasn't clear. My previous comment was about the ordering only, not which actions are performed. If you want to update templates, you still need `--templates`. If you want to apply configuration to other vms (non-templates), then you need `--all`.

Comment (conorsch) [reply dom0/securedrop-update]: That would explain the behavior I'm seeing locally, @marmarek; many thanks for your guidance here!

**Comment by redshiftzero** [comment]
> I've looked into the Journalist updater code, and unfortunately the QtWidget class used by that appears to be PyQt5-only, my local dom0 only has PyQt4

Note we can without too much difficulty migrate PyQt5 -> PyQt4 such that we can reuse that GUI application

**Comment by emkll** [comment]
I've addressed the feedback with the actions performed in `securedrop-update` (update dom0, templates, and AppVMs), and made some changes due to (new) flake8 failures. This now looks good to merge from my pespective. 
Since I was the last one to push to this branch, would either @joshuathayer or @conorsch mind taking one last look at this PR?

**Comment by conorsch** [comment]
Thanks for handling the flake8 fixes, @emkll! Changes LGTM. Let's give @joshuathayer a chance to test prior to merging.

**Comment by joshuathayer** [comment]
Regarding UX for altering users about the need to update: an another option, we could probably add some notification to the client application UI, via a qubes-rpc job from dom0 to sd-svs which would twiddle some state in the securedrop-client DB. 

In terms of review... I've been running the update script now and it's been running at least 2 hours. I've not observed any errors yet but that seems significantly longer than other people's experience, eh?

File: dom0/securedrop-update [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        --icon /usr/share/securedrop/icons/sd-logo.png \
+        --expire-time 60000 \
+        'SecureDrop: $msg'"
+}
+
+# `qubesctl pkg.upgrade` will automatically update dom0 packages, as well,
+# but we *first* want the freshest RPMs from dom0, *then* we'll want to
+# update the VMs themselves.
+securedrop-update-feedback "SecureDrop: Updating dom0 configuration..."
+sudo qubes-dom0-update -y
```
Comment (emkll) [inline dom0/securedrop-update]: adding the `--clean` option here will refresh dnf cache, which might be useful in some cases.

I just had an issue where qubes-dom0-update was complaining of an unsigned package, due to me attempting to download an older whonix template in an effort to reproduce https://github.com/freedomofpress/securedrop-workstation/issues/122#issuecomment-433971413

File: dom0/securedrop-update [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        --app-name 'SecureDrop Workstation' \
+        --icon /usr/share/securedrop/icons/sd-logo.png \
+        --expire-time 60000 \
+        'SecureDrop: $msg'"
+}
+
+# `qubesctl pkg.upgrade` will automatically update dom0 packages, as well,
+# but we *first* want the freshest RPMs from dom0, *then* we'll want to
+# update the VMs themselves.
+securedrop-update-feedback "SecureDrop: Updating dom0 configuration..."
```
Comment (joshuathayer) [inline dom0/securedrop-update]: Minor nit: "SecureDrop:" is added in `securedrop-update-feedback()`, so isn't needed in the message here.

**Comment by conorsch** [comment]
> I eventually manually killed my first run of the update script, which seemed to be hung after running for 3+ hours.

I've seen this happen only once in about a dozen runs—I suspect it's the Whonix VM updates, since those updates often crawl. Let's keep an eye on it over time. Since updates happen in the background, we can count on the updates being installed in a timely fashion—but if Whonix updates require hours, that may require a special fix, e.g. preventing proxying of Whonix updates over Tor.

**Comment by kushaldas** [comment]
>I eventually manually killed my first run of the update script, which seemed to be hung after running for 3+ hours. Rerunning it succeeded in just a few minutes.

This happens most of the time in the `whonix` templates for me, the `apt-get update` even gets stuck (I tried to do that manually) to fetch any file.

**Comment by emkll** [comment]
Some packages are not updating for me after running the updater, seems to be limited to the packages served by the qubes repo. Can anyone reproduce this on a template that was newly updated by the `securedrop-update` script? 
```
$ sudo apt list --upgradable 
Listing... Done
libvchan-xen/unknown 4.0.5-1+deb9u1 amd64 [upgradable from: 4.0.3-1+deb9u1]
libxen-4.8/unknown 2001:4.8.4-5+deb9u1 amd64 [upgradable from: 2001:4.8.4-2+deb9u1]
qubes-gui-agent/unknown 4.0.17-1+deb9u1 amd64 [upgradable from: 4.0.16-1+deb9u1]
xen-utils-common/unknown 2001:4.8.4-5+deb9u1 all [upgradable from: 2001:4.8.4-2+deb9u1]
xserver-xorg-input-qubes/unknown 4.0.17-1+deb9u1 amd64 [upgradable from: 4.0.16-1+deb9u1]
xserver-xorg-video-dummyqbs/unknown 4.0.17-1+deb9u1 amd64 [upgradable from: 4.0.16-1+deb9u1]
```

**Comment by conorsch** [comment]
One more tweak, squashed into prior commits. Tests are passing for me. Take it for a spin, @emkll. 

Details in the commit messages, but looks like the new approach only works against Debian VMs. That's fine for merge, but something we should be aware of.

**Pull request merged**